src/components/particle-emitter.js

```diff
@@ -0,0 +1,147 @@
+import { ParticleEmitter } from "@mozillareality/three-particle-emitter";
+import { textureLoader } from "../utils/media-utils";
+
+AFRAME.registerComponent("particle-emitter", {
+  schema: {
+    src: { type: "string" },
+    startColor: { type: "color" },
+    middleColor: { type: "color" },
+    endColor: { type: "color" },
+    startOpacity: { type: "number" },
+    middleOpacity: { type: "number" },
+    endOpacity: { type: "number" },
+    colorCurve: { type: "string" },
+    sizeCurve: { type: "string" },
+    startSize: { type: "number" },
+    endSize: { type: "number" },
+    sizeRandomness: { type: "number" },
+    ageRandomness: { type: "number" },
+    lifetime: { type: "number" },
+    lifetimeRandomness: { type: "number" },
+    particleCount: { type: "number" },
+    startVelocity: { type: "vec3" },
+    endVelocity: { type: "vec3" },
+    velocityCurve: { type: "string" },
+    angularVelocity: { type: "number" }
+  },
+
+  init() {
+    this.particleEmitter = new ParticleEmitter(null);
+    this.particleEmitter.visible = false;
+    this.el.setObject3D("particle-emitter", this.particleEmitter);
+    this.updateParticles = false;
+  },
+
+  update(prevData) {
+    const data = this.data;
+    const particleEmitter = this.particleEmitter;
+
+    if (prevData.src !== data.src) {
+      this.textureLoaded = false;
+      textureLoader.load(
+        data.src,
+        texture => {
+          this.particleEmitter.material.uniforms.map.value = texture;
+          this.particleEmitter.visible = true;
+          this.updateParticles = true;
+        },
+        undefined,
+        console.error
+      );
+    }
+
+    if (prevData.startColor !== data.startColor) {
+      particleEmitter.startColor.set(data.startColor);
+      this.updateParticles = true;
+    }
+
+    if (prevData.middleColor !== data.middleColor) {
```

Comment:
Unless there are perf concerns (that are worse than doing lots of if checks), the only `if` statements that seem necessary here are the ones that cause you to set `this.updateParticles = true` or load a texture from a new `src`. Perhaps replace the former with

```
if ( predicateA || predicateB || predicateC || predicateD...) {
  this.updateParticles = true;
}
particleEmitter.foo.copy(data.foo);
particleEmitter.bar = data.bar
(etc)
```